pkg/verifier/cosign/cosign.go

```diff
@@ -132,8 +167,157 @@ func (v *cosignVerifier) CanVerify(_ context.Context, referenceDescriptor ocispe
+			if pubKey.ProviderType == azurekeyvault.ProviderName {
+				switch keyType := pubKey.Key.(type) {
+				case *rsa.PublicKey:
+					switch keyType.Size() {
+					case 256:
+						hashType = crypto.SHA256
+					case 384:
+						hashType = crypto.SHA384
+					case 512:
+						hashType = crypto.SHA512
+					default:
+						return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("RSA key check: unsupported key size: %d", keyType.Size())), nil
+					}
+
+					// TODO: remove section after fix for bug in cosign azure key vault implementation
+					// tracking issue: https://github.com/sigstore/sigstore/issues/1384
+					// summary: azure keyvault implementation ASN.1 encodes sig after online signing with keyvault
+					// EC verifiers in cosign have built in ASN.1 decoding, but RSA verifiers do not
+					base64DecodedBytes, err := base64.StdEncoding.DecodeString(blob.Annotations[static.SignatureAnnotationKey])
+					if err != nil {
+						return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("RSA key check: failed to decode base64 signature: %w", err)), nil
+					}
+					// decode ASN.1 signature to raw signature if it is ASN.1 encoded
+					decodedSigBytes, err := decodeASN1Signature(base64DecodedBytes)
+					if err != nil {
+						return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("RSA key check: failed to decode ASN.1 signature: %w", err)), nil
+					}
+					encodedBase64SigBytes := base64.StdEncoding.EncodeToString(decodedSigBytes)
+					sig, err = static.NewSignature(blobBytes, encodedBase64SigBytes, staticOpts...)
+					if err != nil {
+						return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("RSA key check: failed to generate static signature: %w", err)), nil
+					}
+				case *ecdsa.PublicKey:
+					switch keyType.Curve {
+					case elliptic.P256():
+						hashType = crypto.SHA256
+					case elliptic.P384():
+						hashType = crypto.SHA384
+					case elliptic.P521():
+						hashType = crypto.SHA512
+					default:
+						return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("ECDSA key check: unsupported key curve: %s", keyType.Params().Name)), nil
+					}
+				default:
+					return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("unsupported public key type: %T", pubKey)), nil
+				}
+			}
+
+			// return the correct verifier based on public key type and bytes
+			verifier, err := signature.LoadVerifier(pubKey.Key, hashType)
+			if err != nil {
+				return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("failed to load public key from provider [%s] name [%s] version [%s]: %w", mapKey.Provider, mapKey.Name, mapKey.Version, err)), nil
+			}
+			cosignOpts.SigVerifier = verifier
+			// verify signature with cosign options + perform bundle verification
+			bundleVerified, err := cosign.VerifyImageSignature(ctx, sig, subjectDescHash, &cosignOpts)
+			extension := cosignExtension{
+				SignatureDigest: blob.Digest,
```

Comment:
So SignatureDIgest may not actually be a very helpful to distinguish between multiple signatures in a single cosign sig image index. Take an example of an image signed twice by 2 different keys: 
![image](https://github.com/deislabs/ratify/assets/12980641/6cf634f5-4251-4bdc-932d-6890a61389c9)
We can see that the digests are identical of the 2 sig blobs. However, the difference is the sig in the annotation. The digest is the same because the signing payload is the exact same.

Reply:
sounds good to me.

Reply:
I've now divided signature verification by signature. So each sig will have a list of extensions showing the verifications performed each key. 
![image](https://github.com/deislabs/ratify/assets/12980641/ac1d0d32-3a76-4ba8-99fe-8f7f2365bd1e)